lib/perf_hooks.js

```diff
@@ -0,0 +1,311 @@
+'use strict';
+
+const {
+  PerformanceMark,
+  PerformanceMeasure,
+  getEntries,
+  resetMark,
+  clearEntry,
+  nodeTiming,
+  nodeFrame,
+  nodeTimingKeys,
+  observersBuffer,
+  setupObservers,
+  timeOrigin,
+  constants
+} = process.binding('performance');
+
+const {
+  NODE_PERFORMANCE_ENTRY_TYPE_NODE,
+  NODE_PERFORMANCE_ENTRY_TYPE_FRAME,
+  NODE_PERFORMANCE_ENTRY_TYPE_MARK,
+  NODE_PERFORMANCE_ENTRY_TYPE_MEASURE,
+} = constants;
+
+const kInspect = require('internal/util').customInspectSymbol;
+const kCallback = Symbol('callback');
+const kTypes = Symbol('types');
+const kEntries = Symbol('entries');
+const kBuffer = Symbol('buffer');
+const kBuffering = Symbol('buffering');
+const kQueued = Symbol('queued');
+
+const marks = new Map();
+const measures = new Map();
+const observers = new Map();
+const observerCounts = new Uint32Array(observersBuffer);
+const observerableTypes = [ 'node', 'frame', 'mark', 'measure' ];
+
+let errors;
+function lazyErrors() {
+  if (errors === undefined)
+    errors = require('internal/errors');
+  return errors;
+}
+
+function doNotify() {
+  this[kQueued] = false;
+  this[kCallback](this[kBuffer], this);
+  this[kBuffer][kEntries].clear();
+}
+
+// Set up the callback used to receive PerformanceObserver notifications
+function observersCallback(entry) {
+  const type = mapTypes(entry.entryType);
+  const set = getObserversSet(type);
```

Comment:
The set is to ensure that items are only added once. I'm looking in to see if we need to keep that requirement.